### PR TITLE
feat: natgw qos

### LINKED ISSUE #2633: More QoS features supported in NATGW
### Feature request

Trying to implement more QoS features supported in NATGW.

### Use case

The following QoS features will be supported:
* Bandwidth rate limit for NATGW net1 NIC
* Bandwidth rate limit for NATGW net1 EIP (already supported in #2496)
* Bandwidth rate limit for external network traffic on NATGW net1

### PR DESCRIPTION
**Pull request by shane965**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features

### Which issue(s) this PR fixes:
Fixes #2633 
### <samp>🤖 Generated by Copilot at c2af2bf</samp>

This pull request adds QoS policy support for VpcNatGateway objects, which are used to provide NAT services for virtual private clouds. It updates the kube-ovn API, the generated client, the CRD, and the controller logic to enable users to specify and query the QoS policy that is bound to a VpcNatGateway object. It also adds validation, error handling, and synchronization for the QoS policy rules in the OVN database and the pods. It modifies the cluster role of the kube-ovn-controller to allow updating the status of the VpcNatGateway objects.

### <samp>🤖 Generated by Copilot at c2af2bf</samp>

> _The `VpcNatGateway` CRD_
> _Has a new feature, you see_
> _It can use `QoSPolicy`_
> _To limit the traffic spree_
> _And update its status with ease_

### <samp>🤖 Generated by Copilot at c2af2bf</samp>

*  Add the subresource status for the VpcNatGateway CRD to the cluster role of the kube-ovn-controller, which allows the controller to update the status of the VpcNatGateway objects ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-3a12c260f1cdf647b71ad5f0d5caac95cc424e2f31338d15c25c674d49cbf17dR14), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-e8ff83be2329052ff8d94933b1ee843421bd0ebce8f505a4f27fba7f887b90b0R21), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-758698341a22066119ba6e800153bd169d69b0b2017bbf932d5f7cbefac4afadR20))
*  Add a Status field to the VpcNatGateway type, which holds the QoSPolicy name that is applied to the VpcNatGateway object, and a QoSPolicy field to the VpcNatSpec type, which specifies the name of the QoS policy to apply to the VpcNatGateway object ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL482-R484), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL493-R501))
*  Add the DeepCopy and DeepCopyInto methods for the VpcNatStatus type, which are used for copying the VpcNatStatus objects, and a line to the DeepCopyInto method of the VpcNatGateway type, which copies the Status field from the source to the destination object ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-06f8e5bc84406eb8a89c72c345fc4ff5f144e480103ad55d681f7567f0f85515R2332), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-06f8e5bc84406eb8a89c72c345fc4ff5f144e480103ad55d681f7567f0f85515R2422-R2437))
*  Add the UpdateStatus method to the VpcNatGatewayInterface type, which is an interface for the VpcNatGateway client, and the implementation of the UpdateStatus method for the vpcNatGateways type, which is a struct that implements the VpcNatGatewayInterface, and uses the REST client to send a PUT request to update the status of a VpcNatGateway object ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-e02633606c5b7b828d15b3cf16870aff07149b40ef088bc971a53314017d22d9R43), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-e02633606c5b7b828d15b3cf16870aff07149b40ef088bc971a53314017d22d9R132-R146))
*  Add the UpdateStatus method to the FakeVpcNatGateways type, which is a fake client for testing the VpcNatGateway objects, and mimics the behavior of updating the status of a VpcNatGateway object ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-871d234b11e15da1183c728b3c4ade733f29edf57c26cf5407fe21d511aa9a70R99-R109))
*  Add a Bytes method to the VpcNatStatus type, which converts the status to a JSON byte array, and is used for patching the status of the VpcNatGateway objects ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-8e705abac23a0d3a04e999fbbb84ae8907471e95ffc24a9317487f41ee2e11f4R140-R149))
*  Add a new constant QoSBindingTypeNatGw to the QoSPolicyBindingType type, which represents the binding type of a QoS policy to a VpcNatGateway object, and a block of code to the handleUpdateQoSPolicy function, which checks if the QoS policy is not shared and has the binding type of QoSBindingTypeNatGw, which is not supported, and returns an error ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL47-R48), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662L357-R403))
*  Add a block of code to the reconcileEIPBandtithLimitRules function, which checks if the QoS policy bandwidth limit rules have the match type of "ip" and if the match value is a valid CIDR, and returns an error if not ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662R342-R359))
*  Add two constants QoSAdd and QoSDel to the controller package, which represent the operations of adding and deleting the QoS policy for the VpcNatGateway objects, and several functions to the controller package, which are used to execute the QoS policy for the VpcNatGateway objects, such as patchNatGwQoSStatus, execNatGwQoS, execNatGwBandtithLimitRules, execNatGwQoSInPod, and execNatGwRules ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R45-R46), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R911-R1014))
*  Add a function isStatefulSetChanged to the controller package, which is used to compare the fields of two StatefulSet objects and return true if they are different, and a variable needToUpdate to the handleAddOrUpdateVpcNatGw function, which is used to indicate if the StatefulSet object for the VpcNatGateway object needs to be updated ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R224-R252), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R280))
*  Add a block of code to the handleAddOrUpdateVpcNatGw function, which checks if the StatefulSet object for the VpcNatGateway object has changed by calling the isStatefulSetChanged function, and sets the needToUpdate variable to true, and a condition to the same function, which checks if the needToUpdate variable is true before updating the StatefulSet object for the VpcNatGateway object ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4L256-R294), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4L268-R305))
*  Add an else block to the handleAddOrUpdateVpcNatGw function, which checks if the QoSPolicy name in the spec and status of the VpcNatGateway object are different, and calls the execNatGwQoS function to add or delete the QoS policy for the VpcNatGateway object, and then calls the patchNatGwQoSStatus function to update the status of the VpcNatGateway object ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4L275-R334))
*  Add a block of code to the handleInitVpcNatGw function, which checks if the QoSPolicy name in the spec of the VpcNatGateway object is not empty, and calls the execNatGwQoS function to add the QoS policy for the VpcNatGateway object, and then checks if the QoSPolicy name in the spec and status of the VpcNatGateway object are different, and calls the patchNatGwQoSStatus function to update the status of the VpcNatGateway object, and finally calls the updateCrdNatGw function to update the labels of the VpcNatGateway object with the QoSPolicy name ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R381-R401))
*  Modify the signature of the updateCrdNatGw function, which is used to update the labels of the VpcNatGateway object, and add a second argument qos, which represents the QoSPolicy name to update, and a line to the same function, which sets the QoSLabel to the qos argument, and a block of code to the same function, which checks if the QoSLabel of the VpcNatGateway object is different from the qos argument, and sets the op variable to "replace", sets the QoSLabel to the qos argument, and sets the needUpdateLabel variable to true ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4L775-R849), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R864), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R878-R882))
*  Remove a block of code from the handleInitVpcNatGw function, which calls the updateCrdNatGw function to update the labels of the VpcNatGateway object, which is redundant since the handleAddOrUpdateVpcNatGw function will do the same ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4L299-L304))
*  Modify the call to the updateCrdNatGw method in the initSyncCrdVpcNatGw function, which is used to sync the VpcNatGateway objects with the OVN database, and add an empty string as the second argument to the updateCrdNatGw method, which represents the QoSPolicy name to update ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-92aa62435bc2b1b2be34f478f650597068bcf35b8d5975d119f91cf78ac54988L710-R710))
*  Modify the call to the natLabelEip function and the patchEipQoSStatus function in the handleUpdateIptablesEip function, which is used to handle the update events of the EIP objects, and replace the cachedEip.Spec.QoSPolicy argument with the cachedEip.Status.QoSPolicy argument, which represents the QoSPolicy name that is applied to the EIP object ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L414-R419))
*  Add a log message to the enqueueUpdateVpcNatGw function, which is used to enqueue the update events of the VpcNatGateway objects to the work queue, and a log message to the enqueueDeleteVpcNatGw function, which is used to enqueue the delete events of the VpcNatGateway objects to the work queue ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R124), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R135))
*  Add the subresources field to the VpcNatGateway CRD, which enables the status subresource for the VpcNatGateway objects, and the qosPolicy field to the openAPIV3Schema of the VpcNatGateway CRD, which defines the QoSPolicy name for the spec and status of the VpcNatGateway objects ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-7b60edb5fc7c18c120f5a0077606604cfb6be9bb55517095bb9380877ef716c5L168-R178), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-7b60edb5fc7c18c120f5a0077606604cfb6be9bb55517095bb9380877ef716c5R196-R197))
*  Add the regexp package to the `qos_policy.go` file, which is used to validate the format of the CIDR values in the QoS policy bandwidth limit rules, and the reflect package to the `vpc_nat_gateway.go` file, which is used to compare the fields of the StatefulSet objects for the VpcNatGateway objects ([link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662R7), [link](https://github.com/kubeovn/kube-ovn/pull/2753/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4R8))

### CODE REVIEW

File: pkg/controller/vpc_nat_gateway.go [hunk 1]
```diff
@@ -216,6 +221,32 @@ func (c *Controller) handleDelVpcNatGw(key string) error {
...
+		!reflect.DeepEqual(oldSts.Spec.Template.Spec.Containers[0].Command, newSts.Spec.Template.Spec.Containers[0].Command) ||
+		!reflect.DeepEqual(oldSts.Spec.Template.Spec.Containers[0].Args, newSts.Spec.Template.Spec.Containers[0].Args) ||
+		!reflect.DeepEqual(oldSts.Spec.Template.Spec.Containers[0].ImagePullPolicy, newSts.Spec.Template.Spec.Containers[0].ImagePullPolicy) ||
+		!reflect.DeepEqual(oldSts.Spec.Template.Spec.Containers[0].SecurityContext, newSts.Spec.Template.Spec.Containers[0].SecurityContext) ||
+		!reflect.DeepEqual(oldSts.Spec.Template.Spec.NodeSelector, newSts.Spec.Template.Spec.NodeSelector) ||
+		!reflect.DeepEqual(oldSts.Spec.Template.Spec.Tolerations, newSts.Spec.Template.Spec.Tolerations) ||
+		!reflect.DeepEqual(oldSts.Spec.Template.Spec.Affinity, newSts.Spec.Template.Spec.Affinity) ||
+		!reflect.DeepEqual(oldSts.Spec.UpdateStrategy, newSts.Spec.UpdateStrategy) {
+		return true
+	}
```
Comment (shane965) [inline pkg/controller/vpc_nat_gateway.go:245]: make it more readable and more easy to maintain

Comment (bobz965) [reply pkg/controller/vpc_nat_gateway.go:245]: 我感觉这里只需要对照 vpc-nat-gw本身前后是否有关于sts的必要参数发生变化，如果有就重新生成一份sts 更新即可，应该不需要判断sts的变化。sts是基于vpc-nat-gw 产生的

File: pkg/controller/vpc_nat_gateway.go [hunk 2]
```diff
@@ -265,14 +299,36 @@ func (c *Controller) handleAddOrUpdateVpcNatGw(key string) error {
...
+					return err
+				}
+			}
+			if gw.Spec.QoSPolicy != "" {
+				if err = c.execNatGwQoS(gw, gw.Spec.QoSPolicy, QoSAdd); err != nil {
+					klog.Errorf("failed to del qos for natgw %s, %v", key, err)
+					return err
+				}
+			}
+			// if update qos success, will update nat gw status\
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gateway.go:324]:
```suggestion
			// if update qos success, will update nat gw status
```

File: pkg/controller/vpc_nat_gateway.go [hunk 3]
```diff
@@ -265,14 +299,36 @@ func (c *Controller) handleAddOrUpdateVpcNatGw(key string) error {
...
+			if err = c.patchNatGwQoSStatus(key, gw.Spec.QoSPolicy); err != nil {
+				klog.Errorf("failed to patch status for natgw %s, %v", key, err)
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gateway.go:326]:
```suggestion
				klog.Errorf("failed to patch status for nat gw %s, %v", key, err)
```

File: pkg/controller/vpc_nat_gateway.go [hunk 4]
```diff
@@ -325,6 +375,27 @@ func (c *Controller) handleInitVpcNatGw(key string) error {
 		klog.Error(err)
 		return err
 	}
+
+	if gw.Spec.QoSPolicy != "" {
+		if err = c.execNatGwQoS(gw, gw.Spec.QoSPolicy, QoSAdd); err != nil {
+			klog.Errorf("failed to add qos for natgw %s, %v", key, err)
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gateway.go:381]:
```suggestion
			klog.Errorf("failed to add qos for nat gw %s, %v", key, err)
```

File: pkg/controller/vpc_nat_gateway.go [hunk 5]
```diff
@@ -325,6 +375,27 @@ func (c *Controller) handleInitVpcNatGw(key string) error {
...
+			return err
+		}
+	}
+	// if update qos success, will update nat gw status
+	if gw.Spec.QoSPolicy != gw.Status.QoSPolicy {
+		if err = c.patchNatGwQoSStatus(key, gw.Spec.QoSPolicy); err != nil {
+			klog.Errorf("failed to patch status for natgw %s, %v", key, err)
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gateway.go:388]:
```suggestion
			klog.Errorf("failed to patch status for nat gw %s, %v", key, err)
```

File: dist/images/vpcnatgateway/nat-gateway.sh [hunk 6]
```diff
@@ -310,27 +313,104 @@ function eip_egress_qos_add() {
...
+            if [ "$?" -eq 0 ];then
+                delete_tc_u32_filter $dev $qdisc_id $cidr $matchDirection
+            fi
+            exec_cmd "tc filter add dev $dev parent $qdisc_id protocol ip prio $priority u32 match $matchType $matchDirection $cidr police rate "$rate"Mbit burst "$burst"Mb drop flowid :1"
+        elif [ "$classifierType" == "matchall" ];then
+            # del old filter
+            tc -p -s -d filter show dev $dev parent $qdisc_id | grep -w matchall
+            if [ "$?" -eq 0 ];then
+                exec_cmd "tc filter del dev $dev parent $qdisc_id protocol ip prio $priority matchall"
+            fi
```
Comment (bobz965) [inline dist/images/vpcnatgateway/nat-gateway.sh:363]: 仅维护add 逻辑会更好一些，del就用下面的

**Review by bobz965 — CHANGES_REQUESTED** [review]

File: pkg/controller/vpc_nat_gateway.go (hunk 1 above)
Comment (shane965) [reply pkg/controller/vpc_nat_gateway.go:245]: 嗯，我也有想过通过nat-gw status的字段来判断是否有变化，如果这么做需要在冗余很多状态信息到natgw的crd，优点是判断是否变更的函数会更简单一点，确实是crd需要冗余比较多的信息，特别是`affinity`这个字段

File: pkg/controller/qos_policy.go [hunk 7]
```diff
@@ -338,6 +339,24 @@ func (c *Controller) reconcileEIPBandtithLimitRules(
 }
 
 func (c *Controller) validateQosPolicy(qosPolicy *kubeovnv1.QoSPolicy) error {
+	var err error
+	if qosPolicy.Spec.BandwidthLimitRules != nil {
+		for _, rule := range qosPolicy.Spec.BandwidthLimitRules {
+			if rule.MatchType == "ip" {
+				re := regexp.MustCompile(`^(src|dst) (\d{1,3}\.){3}\d{1,3}/\d{1,2}$`)
```
[the remark this answers is not in the document]
Comment (shane965) [reply pkg/controller/qos_policy.go:346]: 单独校验CIDR用现有的net库会更好一点，但是这个cidr前面还有个方向src | dst，感觉用正则可能会更直接点

[the remark this answers is not in the document]
Comment (bobz965) [reply pkg/controller/qos_policy.go:346]: 那不如就拆开 分别校验  再组合作为参数吧

File: pkg/controller/qos_policy.go [hunk 8]
```diff
@@ -337,7 +339,43 @@ func (c *Controller) reconcileEIPBandtithLimitRules(
...
+	}
+
+	direction := parts[0]
+	// invalid direction
+	if direction != "src" && direction != "dst" {
+		return false
+	}
+
+	cidr := parts[1]
+	_, _, err := net.ParseCIDR(cidr)
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:356]: show err near err

File: pkg/controller/vpc_nat_gateway.go [hunk 9]
```diff
@@ -216,6 +221,27 @@ func (c *Controller) handleDelVpcNatGw(key string) error {
 	return nil
 }
 
+func isNatGwStsStatusChanged(gw *kubeovnv1.VpcNatGateway) bool {
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gateway.go:224]:
```suggestion
func isVpcNatGwChanged(gw *kubeovnv1.VpcNatGateway) bool {
```

File: pkg/controller/vpc_nat_gateway.go [hunk 10]
```diff
@@ -264,15 +293,45 @@ func (c *Controller) handleAddOrUpdateVpcNatGw(key string) error {
...
 		return nil
-	} else {
+	} else if needToUpdate {
 		if _, err := c.config.KubeClient.AppsV1().StatefulSets(c.config.PodNamespace).
 			Update(context.Background(), newSts, metav1.UpdateOptions{}); err != nil {
 			err := fmt.Errorf("failed to update statefulset '%s', err: %v", newSts.Name, err)
 			klog.Error(err)
 			return err
 		}
+		if err = c.patchNatGwStsStatus(key); err != nil {
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gateway.go:308]: sts 只是承载 vpcNatGw， 命名上vpcNatGw就可以，不用跟sts那么详细，之前用的deploy，

File: pkg/controller/vpc_nat_gateway.go [hunk 11]
```diff
@@ -828,3 +908,152 @@ func (c *Controller) getNatGw(router, subnet string) (string, error) {
...
+				return nil
+			}
+			klog.Errorf("failed to patch gw %s, %v", gw.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Controller) patchNatGwStsStatus(key string) error {
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gateway.go:947]:
```suggestion
func (c *Controller) patchNatGwStatus(key string) error {
```

**Review by bobz965 — CHANGES_REQUESTED** [review]

**Review by bobz965 — APPROVED** [review]

**Pull request merged**